File: app/src/main/java/com/o4x/musical/imageloader/glide/targets/palette/PaletteTargetListener.java
package com.o4x.musical.imageloader.glide.targets.palette;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.AsyncTask;

import androidx.annotation.Nullable;
import androidx.palette.graphics.Palette;

import com.o4x.musical.helper.MyPalette;

import org.jetbrains.annotations.NotNull;

public abstract class PaletteTargetListener extends AbsPaletteTargetListener {

    private AsyncTask<Bitmap, Void, Palette> paletteAsyncTask;
    private final Context context;

    protected PaletteTargetListener(Context context) {
        this.context = context;
    }

    @Override
    public void onResourceReady(@Nullable Bitmap resource) {
        if (paletteAsyncTask != null) paletteAsyncTask.cancel(false);

        if (resource == null) {
            onColorReady(new MyPalette(context, null), null);
        } else {
            final Palette.Builder paletteBuilder =
                    Palette.from(resource);
            if (isSync) {
                onColorReady(
                        new MyPalette(
                                context,
                                paletteBuilder.generate()
                        ), resource
                );
            } else {
                paletteAsyncTask = paletteBuilder.generate(
                        palette -> onColorReady(new MyPalette(context, palette), resource)
                );
            }
        }
    }

    public abstract void onColorReady(@NotNull MyPalette colors, @Nullable Bitmap resource);
}

File: app/src/main/java/github/o4x/musical/imageloader/glide/targets/palette/PaletteTargetListener.kt
package github.o4x.musical.imageloader.glide.targets.palette

import android.content.Context
import android.graphics.Bitmap
import android.os.AsyncTask
import androidx.palette.graphics.Palette
import github.o4x.musical.helper.MyPalette

abstract class PaletteTargetListener(context: Context) : AbsPaletteTargetListener(context) {

    private var paletteAsyncTask: AsyncTask<Bitmap, Void, Palette>? = null

    override fun onResourceReady(resource: Bitmap?) {
        if (paletteAsyncTask != null) paletteAsyncTask!!.cancel(false)
        if (resource == null) {
            onColorReady(MyPalette(context, null), null)
        } else {
            val paletteBuilder = Palette.from(resource)
            if (isSync) {
                onColorReady(
                    MyPalette(
                        context,
                        paletteBuilder.generate()
                    ), resource
                )
            } else {
                paletteAsyncTask = paletteBuilder.generate { palette: Palette? ->
                    onColorReady(
                        MyPalette(
                            context,
                            palette
                        ), resource
                    )
                }
            }
        }
    }

    abstract fun onColorReady(colors: MyPalette, resource: Bitmap?)
}